uspace/drv/bus/usb/ehci/hw_struct/transfer_descriptor.h

```diff
@@ -74,7 +75,13 @@ typedef struct td {
 
 	/* 64 bit struct only */
 	volatile uint32_t extended_bp[5];
-} td_t;
+
+	/* TDs must be 32-byte aligned */
+	PADD32 [3];
+
+} __attribute__((packed)) td_t;
```

Comment:
Does this need `packed`? This attribute should only be used if the structure or its members are misaligned in memory.

Reply:
Ad PADD32 macro.. this is an example of quite ugly preprocessor magic that I'd rather be avoided.

Reply:
It shall not need it. But it won't do any harm either, and for me `packed` is a kind of hint that this structure is expected to stay exactly as it was written.

Reply:
@jxsvoboda then why it is used everywhere else? :) If you want new code to stop using it, a slight deprecation note would do do the trick. Otherwise i rather like this one, because it shows that the padding must be there and it's not of any interest for reader, because it's not touched by the code. And it's very self-explanatory.

Reply:
First of all, the older code can be and often is wrong too or could suffer from things we didn't know at the time. So avoiding better practices because the old code does not follow them is IMO itself not a good practice.

Secondly, I think when you have a packed structure, it literally tells the compiler that the alignment of the structure is 1 (besides also packing it). There is a different attribute for alignment, which can specify a minimal alignment (if greater than the alignment inferred from the structure's members). But maybe the comment is wrong and should rather say "TDs must be padded to 32-bytes?"

Reply:
It can do harm if you ever allocate the structure yourself, either on stack or via `malloc()`. `packed` means that all members have alignment of 1 byte. The "staying exactly as written" just follows from the normal C struct layout rules, which basically boils down to "preserve order, and add just enough padding before each member to ensure it's aligned; the entire structure has an alignment equal to the maximum alignment of its members".

Reply:
Also, on some architectures, `packed` informs the compiler it must use more expensive misaligned access instructions, or software emulation thereof. I don't suppose that's important for USB stack, but there's a performance hit from using it.

Reply:
I've added the PADD macro to allow better control over structures shared between sw and hw.
the packed attribute is necessary, otherwise a compiler is free to add any padding as long as the addresses of consecutive members are increasing.
If you drop the attribute, you need to guarantee that the offset of each member is correct (static asserts and such).
It's not the best way. afaik at least gcc on arm (used to?) falls back to byte level access which might break semantics of operations that rely on atomicity of word aligned accesses.
In theory these structures should also have aligned attribute, but since they are allocated from dma allocator (which should guarantee alignment) it's not really necessary (it might help with arm gcc byte level access)

Reply:
Just to clarify things: this is a structure that is shared with hardware, thus in principle must be packed.  It must be aligned to 32-byte boundary, too.

1) packed:  The packed attribute shall not be needed, because it is designed to be aligned correctly. Yet the compiler is allowed to put arbitrary padding otherwise, and as such I would stay with attribute in.
2) alignment: As @jvesely said, we allocate space for it manually page-wise. The manual padding was added to aggregate allocations - previously, a page was allocated for every TD. Now it is allocated as `sizeof(td) * count`, and split manually - there comes the manual padding to ensure alignment. Yes, a static assert shall be there, too.

Outcome (for me): do not remove `packed`, but add `aligned(32)` to allow aligned reads. OK?

Reply:
IMO the risk of any compiler adding arbitrary unnecessary padding is very close to zero, and using `static_assert` just to make sure is a vastly better solution than using `packed`.

Reply:
So, after some discussion on the IRC channel, I think I agree that specifying as packed _all_ structures shared between software and hardware (note, this is something else than MMIO) does make a degree of sense. That's all from me.

Reply:
I feel that we come to the conclusion, that using `packed` make sense, but to avoid some downsides, `aligned(32)` is also necessary. I will walk through structures with similar purpose in [uoex]hci and make it consistent, ok? @le-jzr @jermar

Reply:
I am not sure, but don't aligned(32) and packed contradict each other? When you use both, what does alignof() say?

Reply:
No, they don't:
```
typedef struct foo {
	char pad;
	unsigned long long foo;
} __attribute__((packed,aligned(32))) foo_t;

int main() {
	printf("%zu %zu", sizeof(foo_t), &((foo_t *) NULL)->foo); // 32 1
}
```

Reply:
I think it'd be nice to have HW_MEM_STRUCT(size, align) macro, that would hide the exact attributes. It can even add a static assert check suggested by @le-jzr. Both size and alignment are usually stated in the docs, so it's easy to reference the information in a comment.
It can also change per architecture/compiler if need be.

@Aearsis why is sizeof(foo_t) == 32? in the above example. Was it supposed to be alignof() ?

Reply:
I can't see how to make the macro also add an static assert, without being way too magic :)

@jvesely `alignof` is C++ specific, but `sizeof()` here is correct. That was actually the initial motivation - the structure has to be padded to align at the end in order to allow arrays to be created - because array offsets are just syntax sugar for pointer arithmetics, and pointer arithmetics uses size.

Reply:
@Aearsis There is `_Alignof` since C11.